### PR TITLE
Feature/add total member

### PR DESCRIPTION
**Pull request by Kenny-Gin1**
Issue: https://github.com/DXgovernance/dxdao-contracts/issues/129

- Added a `totalHolder` count for our `ERC20SnapshotRep.sol` contract
- Added a mock for testing `internal` functions and simplifying `mint` functions
- Added tests

Was trying to figure out how to test internal functions and simplify minting. The best practice seemed to just make an inheritance contract overriding functions and exposing internal functions. Let me know if this is the way we should do it, and if there are eventual other issues with this approach.

### CODE REVIEW

File: contracts/utils/ERC20/ERC20SnapshotRep.sol [hunk 1]
```diff
@@ -4,11 +4,17 @@ pragma solidity ^0.8.8;
 import "@openzeppelin/contracts-upgradeable/token/ERC20/extensions/ERC20SnapshotUpgradeable.sol";
 import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
 import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/math/SafeMathUpgradeable.sol";
```
Comment (rossneilson) [inline contracts/utils/ERC20/ERC20SnapshotRep.sol:7]: We shouldnt need safemath for 0.8 right?

Comment (AugustoL) [reply contracts/utils/ERC20/ERC20SnapshotRep.sol:7]: I think we dont, but we are still using the openzeppelin libs.

File: contracts/test/ERC20SnapshotRepMock.sol [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+import "@openzeppelin/contracts-upgradeable/token/ERC20/extensions/ERC20SnapshotUpgradeable.sol";
+import "../utils/ERC20/ERC20SnapshotRep.sol";
+
+// mock class using ERC20SnapshotRep
+// @dev We want to expose the internal functions and test them
+contract ERC20SnapshotRepMock is ERC20SnapshotUpgradeable, ERC20SnapshotRep {
+    constructor() {}
+
+    function mint(address to, uint256 amount) public override {
```
Comment (AugustoL) [inline contracts/test/ERC20SnapshotRepMock.sol:11]: This funcion is not needed, the mint are burned are defined in the ERC20SnapshotRep for only owner

File: contracts/test/ERC20SnapshotRepMock.sol [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+    function _addHolders(address account) public returns (bool) {
+        return addHolders(account);
+    }
+
+    function _removeHolders(address account) public returns (bool) {
+        return removeHolders(account);
+    }
+
+    function _beforeTokenTransfer(
```
Comment (AugustoL) [inline contracts/test/ERC20SnapshotRepMock.sol:23]: looks like thsi fixes a wront use of _beforeTokenTransfer in ERC20SnapshotRep, where the _beforeTokenTransfer uses onlyOwner.

Just remove the 
```
    function _beforeTokenTransfer(
        address from,
        address to,
        uint256 amount
    ) internal virtual override onlyOwner {}
    ``` in the ERC20SnapshotRep and you should be able to remove this _beforeTokenTransfer and everything should work.

**Comment by Kenny-Gin1** [comment]
@rossneilson @AugustoL Thanks for your feedback I've now fixed it in regards to that.

- Changed from plural to singular regarding `addHolder` and `removeHolder` functions
- Removed the `mint` and `_beforeTokenTransfer` functions
- Changed the tests to better reflect usage
- Removed `SafeMath` library since Solidity checks for overflows by default in version 0.8 <

**Comment by AugustoL** [comment]
Also, does all tests run succesfully? the test actions failed https://github.com/DXgovernance/dxdao-contracts/runs/6162734754?check_suite_focus=true

**Comment by Kenny-Gin1** [comment]
> Also, does all tests run succesfully? the test actions failed https://github.com/DXgovernance/dxdao-contracts/runs/6162734754?check_suite_focus=true

Yes! I will look at it together today with @MiltonTulli before merging. Because I think it's in regards to calling the `mint` in `ERC20SnapshotRep`which has the `_snapshot()` function in it.

**Review by AugustoL — APPROVED** [review]

**Comment by AugustoL** [comment]
one tests in failing in the CI

**Review by AugustoL — CHANGES_REQUESTED** [review]
@Kenny-Gin1 try merging with develop, fix tests and commit

**Comment by Kenny-Gin1** [comment]
@AugustoL @MiltonTulli realised that removing this part:     

```
function _beforeTokenTransfer(
        address from,
        address to,
        uint256 amount
    ) internal virtual override onlyOwner {}
```

from the `ERC20SnapshotRep` made one of @MiltonTulli tests fail. Should it still be removed? Are we testing it wrongly for some reason?

**Comment by Kenny-Gin1** [comment]
Alright, all tests are working ✅  @rossneilson @AugustoL

File: package.json [hunk 4]
```diff
@@ -102,8 +102,7 @@
     "*.{ts,tsx,js,jsx,json,md,sol}": [
       "yarn format",
       "yarn solidity-linter",
-      "git add",
```
Comment (AugustoL) [inline package.json]: remove git add too please.

After commit if you have changes left you can add them with `git add -p` and `git commit --amend`

**Review by AugustoL — CHANGES_REQUESTED** [review]

**Review by AugustoL — APPROVED** [review]

File: test/utils/ERC20/ERC20SnapshotRep.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+contract("ERC20SnapshotRep", accounts => {
+  let ERC20SnapshotRep;
+
+  beforeEach(async () => {
+    ERC20SnapshotRep = await ERC20SnapshotRepMock.new({
+      from: accounts[0],
+    });
+    await ERC20SnapshotRep.initialize("DXdao", "DXD");
+    await ERC20SnapshotRep._addHolder(SOME_ADDRESS, { from: accounts[0] });
```
Comment (MiltonTulli) [inline test/utils/ERC20/ERC20SnapshotRep.js:17]: @Kenny-Gin1 I think that you can omit `, {from : accounts[0]}` in all contract function executions. By default will execute with the owner (L.14). 

```
await ERC20SnapshotRep._addHolder(SOME_ADDRESS);
await ERC20SnapshotRep._removeHolder(SOME_ADDRESS);
await ERC20SnapshotRep.mint(SOME_ADDRESS, 100);
await ERC20SnapshotRep.burn(SOME_ADDRESS, 50);
```

**Review by MiltonTulli — APPROVED** [review]

**Pull request merged**